src/main/java/pl/filipdawidczyk/application/service/TransactionService.java

```diff
@@ -0,0 +1,71 @@
+package pl.filipdawidczyk.application.service;
+
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+import pl.filipdawidczyk.application.client.NbpApiClient;
+import pl.filipdawidczyk.application.exception.MissingDataException;
+import pl.filipdawidczyk.application.exception.WrongRequestException;
+import pl.filipdawidczyk.application.model.NbpCurrencyResponse;
+import pl.filipdawidczyk.application.model.Rate;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+import java.util.regex.Pattern;
+
+@AllArgsConstructor
+@Service
+public class TransactionService {
+
+    private final NbpApiClient nbpApiClient;
+
+    public BigDecimal calculate(final String code, final BigDecimal initialAmount, final LocalDate buyDate, final LocalDate sellDate){
+
+        validateRequest(code, initialAmount, buyDate, sellDate);
```

Comment:
request should be validated in separate validator, that is triggered before thread enter the controller